gtk4/src/builder_rust_scope.rs

```diff
@@ -0,0 +1,143 @@
+use crate::subclass::builder_scope::BuilderScopeImpl;
+use crate::Builder;
+use crate::BuilderClosureFlags;
+use crate::BuilderError;
+use crate::BuilderScope;
+use glib::prelude::*;
+use glib::subclass::prelude::*;
+use glib::translate::*;
+use glib::Closure;
+use std::collections::HashMap;
+use std::rc::Rc;
+
+glib::wrapper! {
+    pub struct BuilderRustScope(ObjectSubclass<BuilderRustScopePrivate>)
+        @implements BuilderScope;
+}
+
+#[derive(Default)]
+pub struct BuilderRustScopePrivate {
+    callbacks:
+        std::cell::RefCell<HashMap<String, Rc<dyn Fn(&[glib::Value]) -> Option<glib::Value>>>>,
+}
+
+#[glib::object_subclass]
+impl ObjectSubclass for BuilderRustScopePrivate {
+    const NAME: &'static str = "GtkBuilderRustScope";
+    type Type = BuilderRustScope;
+    type ParentType = glib::Object;
+    type Interfaces = (BuilderScope,);
+}
+
+impl ObjectImpl for BuilderRustScopePrivate {}
+
+impl BuilderScopeImpl for BuilderRustScopePrivate {
+    fn create_closure(
+        &self,
+        _builder_scope: &Self::Type,
+        builder: &Builder,
+        function_name: &str,
+        flags: BuilderClosureFlags,
+        object: Option<&glib::Object>,
+    ) -> Result<Closure, glib::Error> {
+        self.callbacks
+            .borrow()
+            .get(function_name)
+            .ok_or_else(|| {
+                glib::Error::new(
+                    BuilderError::InvalidFunction,
+                    &format!("No function named `{}`", function_name),
+                )
+            })
+            .map(|callback| {
+                let callback = callback.clone();
+                let swapped = flags.contains(BuilderClosureFlags::SWAPPED);
+                let object = object.cloned().or_else(|| builder.current_object());
+                if let Some(object) = object {
+                    let object_weak = object.downgrade();
+                    let closure = if swapped {
+                        Closure::new_local(move |args| {
+                            let mut args = args.to_owned();
+                            let object = object_weak
+                                .upgrade()
+                                .expect("Object destroyed before closure invalidated");
```

Comment:
Maybe that should just cause the closure to return silently as if it was invalidated? But it can't really happen so I guess this is also fine

Reply:
I think this would not ever happen unless there was a serious internal bug in glib or in gtk-rs-core, so that's why I had it panic, maybe there should be a note there that this is an internal assert?